### ISSUE TITLE
possible to show ignored/"preprocessed" files and folders - permissions problems. Missing some dupes in WSL Ubuntu

### ISSUE THREAD

**Issue by james-cook** [sahib/rmlint#428]
I'm using a freshly compiled 2.10.1 on ubuntu(18.04) in WSL windows 10(!) and it works fine :)

The only thing is it seems to miss **some** (not all) duplicates.
In the current case 11,195 files / 209GB are "missed" in a compare of two folders (total size 750GB).

I am "debugging" this now:
1. rmlint reports it is ignoring 3356 files - is there any way to see which files these are?
2. rmlint then "preprocesses" and goes from 107967 usable files to 46343 files - is there any way to show the files weeded out here?

My hunch is that some files/folders have "legacy" permissions (the files and folders are partially on old external HDDs from OTHER windows machines) and may not even be being traversed. 

Does rmlint report permissions issues at all? (I haven't seen this so far)
Is it advised to "sudo" rmlint in such set ups?

OTOH I am using "beyond compare 4" to compare results and this program shows me the 11,195 matches without sudoing etc. (i.e. the files are all readable if not writable).
Beyond compare is flexible about file timestamps - is that a problem with rmlint? - i.e. is the match "size" then "hash" or is it "timestamp" >"size" > "hash" (if timestamps are significant is there an option to "ignore" them?)

Any hints welcome

**Comment by james-cook** [comment]
I change permissions and reran the rmlint command without sudo. This time the 11000 or so files were picked up.
My observations are:

Ignored and preprocessing files were the same as before - so have nothing to do with "my problem". AFAICT ignored just filters out system hidden files and preprocessing probably sets all files without the same size to one side (my assumption).
So it might be useful to "see" these files in some way - but was not responsible for my problem.

I haven't fully checked why the initial runs missed the dupes (due to time constraints) - it could be something other than permissions issues... 
I did not try the last run with sudo - it found all dupes in standard mode (after the permissions changes)

From my side I'll keep this issue open for a time - just to see how my extensive use of rmlint in the next days affects my thinking on this problem.
At the moment I think it would be good if rmlint or probably better a command beforehand would report potential permissions issues.
How this would work I am not sure. I don't currently know if WSL Ubuntu has any knowledge of the "under or over"lying NTFS permissions.

**Comment by sahib** [comment]
> Beyond compare is flexible about file timestamps - is that a problem with rmlint? - i.e. is the match "size" then "hash" or is it "timestamp" >"size" > "hash" (if timestamps are significant is there an option to "ignore" them?)

Timestamps are not taken into account.

> Does rmlint report permissions issues at all? (I haven't seen this so far)
> Is it advised to "sudo" rmlint in such set ups?

Only if it cannot read a file or directory. But it will warn you about this.
And, no ``sudo`` is most definitely a bad idea - not sure if it does matter a lot in case of WSL, but ``rmlint`` does not need it.

> rmlint reports it is ignoring 3356 files - is there any way to see which files these are?

Not really. You can try with ``rmlint -vvv`` to see debug logs, but those include only general hints and not individual files. Might be still useful though.

> rmlint then "preprocesses" and goes from 107967 usable files to 46343 files - is there any way to show the files weeded out here?

This will be mostly files that have unique sizes and some others, depending on your options.

> At the moment I think it would be good if rmlint or probably better a command beforehand would report potential permissions issues.

With what command did you run ``rmlint``? Bad permissions might be filtered out beforehand, but will be printed separately.

> OTOH I am using "beyond compare 4" to compare results and this program shows me the 11,195 matches without sudoing etc. (i.e. the files are all readable if not writable).

Just as a note: different duplicate finders might report different numbers, depending on whether they count hard or symbolic links, filter out path duplicates and pre-process their set of files differently. See also:

* https://rmlint.readthedocs.io/en/latest/faq.html#rmlint-finds-more-less-dupes-than-tool-x
* https://rmlint.readthedocs.io/en/latest/cautions.html

**Comment by james-cook** [comment]
Thanks for your reply.
I can happily report that my issue seems to coming from some where else(!)
I'm using a laptop (W10) with 2 usb hubs on the 2 usb ports.

In the latest run - 1.3 million files, 10 ext. drives and ca. 1.2TB data **I noticed that after 2 or so days one drive was not only suspended but _completely shutdown_**! I had noticed this before, last week, on the other, newer usb hub with a different drive (I tried reconnecting the drive but WSL Ubuntu only recognises drives plugged in before the initial bash is started.)
So what seems to be happening is that **rmlint** continues silently. But some files from this drive are then not in the results.
This is a possible explanation for the missing files.

In addition the laptop rebooted for an update overnight (despite setting updates with user confirmation only) so I was a bit worried **rmlint.sh** might be empty after a 3 day run!

Luckily **rmlint.sh** does output and flush to **rmlint.sh** so I was able to check and run the **rmlint** job despite the original **rmlint** run not completing due to the reboot (obviously rmlint.sh was not complete and I need to rerun the original job later on).
If you are using WSL Ubuntu etc. it pays to check if the drives really are still reachable under `/mnt/X` etc. (you may need to reboot!) before running **rmlint.sh**.
What I also did do was change `return 1` to `exit 1` in `original_check` for the 2 tests original / duplicate have disappeared.
This makes correcting the problem possible if it exists - and makes it easier to re-run rmlint.sh without warnings later on. 

So to recap: for WSL Ubuntu only and from my experience only:

1. Check your drives are not disappearing from WSL Ubuntu during the run of **rmlint.sh**
2. if drives have disappeared just reboot before you run **rmlint.sh**
3. if your windows machine updates overnight all is not lost - run the incomplete **rmlint.sh** as usual
4. perform the edit I mentioned above to **rmlint.sh** just to make extra sure
5. run rmlint.sh to pick up any files missing from the first run

Thanks for the very versatile tool.

**Comment by sahib** [comment]
Okay, good to know. :smile: 

I will close this issue then.

> Thanks for the very versatile tool.
> What I particularly value is the use of hash caching (something e.g. dupeguru doesn't do yet).
> I can confirm that with WSL Ubuntu does set the xattr (extra attributes) when the --xattr param is set.

Thanks for the experience report!
